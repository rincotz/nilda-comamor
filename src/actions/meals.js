import db, { storage, auth, firebase } from '../firebase/firebase'
import {ADD_MEAL, EDIT_MEAL, REMOVE_MEAL, SET_MEALS} from "./constants";

// ADD_MEAL
export const addMeal = (meal) => ({
    type: ADD_MEAL,
    meal
})

export const startAddMeal = (mealData = {}) => {
    return (dispatch) => {
        const {
            name,
            description,
            price,
            available,
            location,
            chef,
            courrier,
            courrierStart,
            courrierEnd,
            pickUp,
            pickUpStart,
            pickUpEnd,
            table,
            tableStart,
            tableEnd,
            frozen,
            picture
        } = mealData
        const meal = {
            name,
            description,
            price,
            available,
            location,
            chef,
            courrier,
            courrierStart,
            courrierEnd,
            pickUp,
            pickUpStart,
            pickUpEnd,
            table,
            tableStart,
            tableEnd,
            frozen,
            picture
        }

        return db.collection('meals')
            .add(meal)
            .then((docRef) => {
                dispatch(addMeal({
                    id: docRef.id,
                    ...meal
                }))
            })
    }
}

export const getCooker = () => {
    auth.onAuthStateChanged(cooker => {
        return db.collection('users').doc(cooker.uid)
    })
}

export const addMealPic = (uid, blob, mealImage) => {
    const storageRef = firebase.storage().ref()
    storageRef.child(`meals/${uid}`).put(blob, { contentType: mealImage.type }).then(snapshot => {
        return snapshot.ref.getDownloadURL()
    })
}

// EDIT_MEAL
export const editMeal = (id, updates) => ({
    type: EDIT_MEAL,
    id,
    updates
})

export const startEditMeal = (id, updates) => {
    return (dispatch) => {
        return db.collection('meals').doc(id).update(updates)
            .then(() => {
                dispatch(editMeal(id, updates))
            })
    }
}

// REMOVE_MEAL
export const removeMeal = ({ id }) => ({
    type: REMOVE_MEAL,
    id
})

export const startRemoveMeal = ({ id }) => {
    return (dispatch) => {
        return db.collection('meals').doc(id).delete()
            .then(dispatch(removeMeal({ id })))
    }
}

// SET_MEALS
export const setMeals = (meals) => ({
    type: SET_MEALS,
    meals
})

export const startSetMeals = () => {
    return (dispatch) => {
        return db.collection('meals')
            .get()
            .then((querySnapshot) => {
                const meals = []
                querySnapshot.forEach((doc) => {
                    meals.push({
                        id: doc.id,
                        ...doc.data()
                    })
                })
                dispatch(setMeals(meals))
            })
    }
}