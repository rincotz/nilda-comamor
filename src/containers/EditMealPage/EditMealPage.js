import React from 'react';
import MealForm from "../MealForm/MealForm";
import { startEditMeal, startRemoveMeal } from "../../actions/meals";
import { connect } from "react-redux";

export class EditMealPage extends React.Component {
    onSubmit = (meal) => {
        this.props.startEditMeal(this.props.meal.id, meal)
        this.props.history.push('/')
    }
    onRemove = () => {
        this.props.startRemoveMeal({ id: this.props.meal.id })
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
                <MealForm
                    onSubmit={this.onSubmit}
                    meal={this.props.meal}
                />
                <button
                    onClick={this.onRemove}
                >
                    Remover
                </button>
            </div>
        )
    }
}

const mapStateToProps = (state, props) => ({
    meal: state.meals.find((meal) => meal.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch, props) => ({
    startEditMeal: (id, meal) => dispatch(startEditMeal(id, meal)),
    startRemoveMeal: (data) => dispatch(startRemoveMeal(data))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditMealPage)