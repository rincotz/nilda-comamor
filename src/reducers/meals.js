import {ADD_MEAL, EDIT_MEAL, REMOVE_MEAL, SET_MEALS} from "../actions/constants";

// Meals Reducer
export const mealDefaultState = {
    id: "",
    available: "",
    chef: {
        uid: "",
        id: "",
        name: "",
        gender: "",
        birth: "",
        email: "",
        phone: "",
        bio: "",
        nationality: "",
        addressLine1: "",
        addressLine2: "",
        number: "",
        neighborhood: "",
        zip: "",
        picture: "",
        geoPoint: "",
        guestRatings: "",
        cookerRatings: "",
        ratings: ""
    },
    courrier: "",
    courrierEnd: "",
    courrierStart: "",
    description: "",
    frozen: "",
    name: "",
    pickUp: "",
    pickUpEnd: "",
    pickUpStart: "",
    picture: "",
    price: "",
    table: "",
    tableEnd: "",
    tableStart: ""
}
const mealsReducerDefaultState = []
export default (state = mealsReducerDefaultState, action) => {
    switch (action.type) {
        case ADD_MEAL:
            return [...state, action.meal]
        case REMOVE_MEAL:
            return state.filter(({ id }) => (
                id !== action.id
            ))
        case EDIT_MEAL:
            return state.map((meal) => {
                if (meal.id === action.id) {
                    return {
                        ...meal,
                        ...action.updates
                    }
                } else {
                    return meal
                }
            })
        case SET_MEALS:
            return action.meals
        default:
            return state
    }
}