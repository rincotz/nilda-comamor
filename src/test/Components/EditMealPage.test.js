import React from 'react'
import { shallow } from 'enzyme'
import { EditMealPage } from "../../containers/EditMealPage/EditMealPage"
import meals from '../fixtures/meals'

let startEditMeal, startRemoveMeal, history, wrapper

beforeEach(() => {
    startEditMeal = jest.fn()
    startRemoveMeal = jest.fn()
    history = { push: jest.fn() }
    wrapper = shallow(
        <EditMealPage
            meal={meals[2]}
            startEditMeal={startEditMeal}
            startRemoveMeal={startRemoveMeal}
            history={history}
        />
    )
})

test('should render EditMealPage correctly', () => {
    expect(wrapper).toMatchSnapshot()
})

// test('should handle editing meals', () => {
//     wrapper.find('MealForm').prop('onSubmit')(meals[2])
//     expect(history.push).toHaveBeenLastCalledWith('/')
//     expect(editMeal).toHaveBeenLastCalledWith(meals[2].id, meals[2])
// })

// test('should handle startRemoveMeal', () => {
//     wrapper.find('button').simulate('click')
//     expect(history.push).toHaveBeenLastCalledWith('/')
//     expect(startRemoveMeal).toHaveBeenLastCalledWith({ id: meals[2].id })
// })