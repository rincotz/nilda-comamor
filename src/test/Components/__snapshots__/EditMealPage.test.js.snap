// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditMealPage correctly 1`] = `
<div>
  <MealForm
    meal={
      Object {
        "available": 7,
        "courrier": true,
        "courrierEnd": 0,
        "courrierStart": 0,
        "description": "espaghetti ao molho de tomates",
        "frozen": false,
        "id": "3",
        "location": 2,
        "name": "macarronada",
        "pickUp": false,
        "pickUpEnd": undefined,
        "pickUpStart": undefined,
        "picture": "",
        "price": 800,
        "table": false,
        "tableEnd": undefined,
        "tableStart": undefined,
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remover
  </button>
</div>
`;
